p2p/transport/webrtc/transport.go

```diff
@@ -0,0 +1,539 @@
+		return nil, fmt.Errorf("listener could not resolve udp address: %w", err)
+	}
+
+	socket, err := net.ListenUDP(nw, udpAddr)
+	if err != nil {
+		return nil, fmt.Errorf("listen on udp: %w", err)
+	}
+
+	listener, err := t.listenSocket(socket)
+	if err != nil {
+		socket.Close()
+		return nil, err
+	}
+	return listener, nil
+}
+
+func (t *WebRTCTransport) listenSocket(socket *net.UDPConn) (tpt.Listener, error) {
+	listenerMultiaddr, err := manet.FromNetAddr(socket.LocalAddr())
+	if err != nil {
+		return nil, err
+	}
+
+	listenerFingerprint, err := t.getCertificateFingerprint()
+	if err != nil {
+		return nil, err
+	}
+
+	encodedLocalFingerprint, err := internal.EncodeDTLSFingerprint(listenerFingerprint)
+	if err != nil {
+		return nil, err
+	}
+
+	certMultiaddress, err := ma.NewMultiaddr(fmt.Sprintf("/webrtc/certhash/%s", encodedLocalFingerprint))
+	if err != nil {
+		return nil, err
+	}
+
+	listenerMultiaddr = listenerMultiaddr.Encapsulate(certMultiaddress)
+
+	return newListener(
+		t,
+		listenerMultiaddr,
+		socket,
+		t.webrtcConfig,
+	)
+}
+
+func (t *WebRTCTransport) Dial(ctx context.Context, remoteMultiaddr ma.Multiaddr, p peer.ID) (tpt.CapableConn, error) {
+	scope, err := t.rcmgr.OpenConnection(network.DirOutbound, false, remoteMultiaddr)
+	if err != nil {
+		return nil, err
+	}
+	err = scope.SetPeer(p)
+	if err != nil {
+		scope.Done()
+		return nil, err
+	}
+	conn, err := t.dial(ctx, scope, remoteMultiaddr, p)
+	if err != nil {
+		scope.Done()
+		return nil, err
+	}
+	return conn, nil
+}
+
+func (t *WebRTCTransport) dial(
+	ctx context.Context,
+	scope network.ConnManagementScope,
+	remoteMultiaddr ma.Multiaddr,
+	p peer.ID,
+) (tConn tpt.CapableConn, err error) {
+	var pc *webrtc.PeerConnection
+	defer func() {
+		if err != nil {
+			if pc != nil {
+				_ = pc.Close()
+			}
+			if tConn != nil {
+				_ = tConn.Close()
+			}
+		}
+	}()
+
+	remoteMultihash, err := internal.DecodeRemoteFingerprint(remoteMultiaddr)
+	if err != nil {
+		return nil, fmt.Errorf("decode fingerprint: %w", err)
+	}
+	remoteHashFunction, ok := internal.GetSupportedSDPHash(remoteMultihash.Code)
+	if !ok {
+		return nil, fmt.Errorf("unsupported hash function: %w", nil)
+	}
+
+	rnw, rhost, err := manet.DialArgs(remoteMultiaddr)
+	if err != nil {
+		return nil, fmt.Errorf("generate dial args: %w", err)
+	}
+
+	raddr, err := net.ResolveUDPAddr(rnw, rhost)
+	if err != nil {
+		return nil, fmt.Errorf("resolve udp address: %w", err)
+	}
+
+	// Instead of encoding the local fingerprint we
+	// generate a random UUID as the connection ufrag.
+	// The only requirement here is that the ufrag and password
+	// must be equal, which will allow the server to determine
+	// the password using the STUN message.
+	ufrag := genUfrag()
+
+	settingEngine := webrtc.SettingEngine{}
+	// suppress pion logs
+	loggerFactory := pionlogger.NewDefaultLoggerFactory()
+	loggerFactory.DefaultLogLevel = pionlogger.LogLevelDisabled
+	settingEngine.LoggerFactory = loggerFactory
+
```

Comment:
nit: @GlenDC Can we add the IncludeLoopbackCandidate setting here?

Reply:
Can you be more specific @ckousik ?

Reply:
Ok added it here @ckousik : https://github.com/libp2p/go-libp2p/pull/1999/commits/99ba632bbaafdf9ce41f27a5f7a9bde1bad357ab
Do unresolve it if you meant something different or if you wanted it to be configurable / conditional.
For now it is hardcoded as `true`.